### PR TITLE
python310Packages.fairseq: init at 0.12.3

### PR DESCRIPTION
**Pull request by happysalada**
###### Description of changes

I'm opening this PR without a linux machine. I'm leaving this open to get ofborg to build this for me so I can move forward until I get my linux machine back.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@natsukium I don't know if I can ask you this, but would you mind helping me on this PR. I need a linux machine to continue the evaluation, and mine is out of service at the moment.
I would be happy to help you on any other PR where you need a review / test with darwin.

File: pkgs/development/python-modules/fairseq/default.nix [hunk 1]
```diff
@@ -0,0 +1,88 @@
+{ lib
+, buildPythonPackage
+, pythonOlder
+, fetchFromGitHub
+
+# Native build inputs
+, cython
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:7]:
```suggestion
, cython
, pythonRelaxDepsHook
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+, regex
+, torch
+, tqdm
+, bitarray
+, torchaudio
+, scikit-learn
+, packaging
+
+# Check inputs
+, pytestCheckHook
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:25]:
```suggestion
, expecttest
, hypothesis
, pytestCheckHook
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+in
+buildPythonPackage rec {
+  inherit version pname;
+
+  # Pypi source package doesn't contain tests
+  src = fetchFromGitHub {
+    owner = "pytorch";
+    repo = pname;
+    rev = "v${version}";
+    hash = "";
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:40]:
```suggestion
    hash = "sha256-XX/grU5ljQCwx33miGoFc/7Uj9fZDtmhm4Fz7L4U+Bc=";
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  };
+
+  disabled = pythonOlder "3.7";
+
+  nativeBuildInputs = [
+    cython
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:46]:
```suggestion
    cython
    pythonRelaxDepsHook
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    which
+  ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:49]:
```suggestion
  pythonRelaxDeps = [
    "hydra-core"
    "omegaconf"
  ];
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    bitarray
+    torchaudio
+    scikit-learn
+    packaging
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    # apex
+  ];
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:68]:
```suggestion
  nativeCheckInputs = [
    expecttest
    hypothesis
    pytestCheckHook
  ];
```

File: pkgs/development/python-modules/fairseq/default.nix [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  pythonImportsCheck = [ "fairseq" ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:71]:
```suggestion
  preCheck = ''
    export HOME=$TMPDIR
    cd tests
  '';
  
  pytestFlagsArray = [
    "--import-mode append"
  ];
```
The pytest used in the current `pytestCheckHook` cannot find the C extension module.

File: pkgs/development/python-modules/fairseq/default.nix [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  disabledTestPaths = [
+    # FIXME: these megatron tests import the optional nvidia apex package
+    "fairseq/model_parallel/megatron/mpu/tests/test_cross_entropy.py"
+    "fairseq/model_parallel/megatron/mpu/tests/test_data.py"
+    "fairseq/model_parallel/megatron/mpu/tests/test_initialize.py"
+    "fairseq/model_parallel/megatron/mpu/tests/test_layers.py"
+    "fairseq/model_parallel/megatron/mpu/tests/test_random.py"
+  ];
```
Comment (natsukium) [inline pkgs/development/python-modules/fairseq/default.nix:79]:
```suggestion
  disabledTests = [
    # this test requires xformers
    "test_xformers_single_forward_parity"
    # this test requires iopath
    "test_file_io_async"
    # these tests require network access
    "test_s2s_transformer_checkpoint"
    "test_librispeech_s2t_transformer_s_checkpoint"
    "test_s2s_transformer_checkpoint"
    "test_waitk_checkpoint"
    "test_sotasty_es_en_600m_checkpoint"
    "test_librispeech_s2t_conformer_s_checkpoint"
  ];
```

**Review by natsukium — COMMENTED** [review]
We can build this package with these patches.

**Comment by happysalada** [comment]
I tried to pick all your commits to retain attribution.
However there was a formatting mistake, when I squashed on my local, it looks like the attribution is gone. :-(

**Comment by natsukium** [comment]
I don't mind that. Thank you. :)

**Comment by happysalada** [comment]
@natsukium ofborg is passing, I think this is ready. Let me know what you think when you have a moment.

**Review by natsukium — APPROVED** [review]
nits: I think redundant comments can be deleted.

**Pull request merged**